### PR TITLE
feat(toHaveScreenshot): enhance messaging in case of failures

### LINKED ISSUE #12563: [BUG] toHaveScreenshot: more information in call log when screenshots differ
See https://playwright.dev/docs/next/test-assertions#custom-expect-message

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch:
- adds call logs to track screenshot timeouts, e.g. due to
  waiting for web fonts
- makes sure all snapshot expectations have `.png` extension
- throws a polite error when given a buffer or a string instead of a
  page or a locator
- removes stray NL between error description and call log

image

Fixes #12563

### CODE REVIEW

File: packages/playwright-core/src/server/screenshotter.ts [hunk 1]
```diff
@@ -138,6 +140,10 @@ export class Screenshotter {
   }
 
   async _preparePageForScreenshot(progress: Progress, disableAnimations: boolean, waitForFonts: boolean) {
+    if (disableAnimations)
+      progress.log('  disabled all CSS animations because "animations" option is set to "disabled"');
```
Comment (dgozman) [inline packages/playwright-core/src/server/screenshotter.ts:144]: `"disabled"` does not play well with languages - let's drop it? Same for `"ready"`.

File: packages/playwright-test/src/expect.ts [hunk 2]
```diff
@@ -257,6 +257,7 @@ function wrap(matcherName: string, matcher: any) {
       reportStepError(e);
     }
   };
+  Object.defineProperty(matcher, 'name', { value: '__PWTRAP__[expect.' + matcherName + ']' });
```
Comment (dgozman) [inline packages/playwright-test/src/expect.ts:260]: Why this change?

File: packages/playwright-test/src/matchers/toMatchSnapshot.ts [hunk 3]
```diff
@@ -203,11 +203,12 @@ class SnapshotHelper<T extends ImageComparatorOptions> {
     if (diffError) {
       output.push(...[
```
Comment (dgozman) [inline packages/playwright-test/src/matchers/toMatchSnapshot.ts:204]:
```suggestion
      output.push(indent(diffError, '  '));
```

File: packages/playwright-test/src/matchers/toMatchSnapshot.ts [hunk 4]
```diff
@@ -289,6 +290,10 @@ export async function toHaveScreenshot(
       testInfo, testInfo._screenshotPath.bind(testInfo), 'png',
       testInfo.project.expect?.toHaveScreenshot || {},
       nameOrOptions, optOptions);
+  if (!helper.snapshotPath.toLowerCase().endsWith('.png'))
+    throw new Error(`Screenshot name "${path.basename(helper.snapshotPath)}" must have '.png' extension`);
+  if (Buffer.isBuffer(pageOrLocator) || (typeof pageOrLocator === 'string'))
```
Comment (dgozman) [inline packages/playwright-test/src/matchers/toMatchSnapshot.ts:295]: There is an `expectType` helper in `playwright-test/src/util.ts`.

File: packages/playwright-core/src/server/page.ts [hunk 5]
```diff
@@ -458,16 +458,24 @@ export class Page extends SdkObject {
...
     } | undefined = undefined;
+    const callTimeout = this._timeoutSettings.timeout(options);
     return controller.run(async progress => {
       let actual: Buffer | undefined;
       let previous: Buffer | undefined;
       const pollIntervals = [0, 100, 250, 500];
-      while (true) {
+      progress.log(`${metadata.apiName}${callTimeout ? ` with timeout ${callTimeout}ms` : ''}`);
+      if (isGeneratingNewScreenshot)
+        progress.log(`  generating new screenshot expectation: waiting for 2 consecutive screenshots to match.`);
```
Comment (dgozman) [inline packages/playwright-core/src/server/page.ts:468]: This is the only line that has a full stop at the end :)

File: packages/playwright-core/src/server/page.ts [hunk 6]
```diff
@@ -458,16 +458,24 @@ export class Page extends SdkObject {
...
+      else
+        progress.log(`  waiting for screenshot to match expectation`);
+      for (let iteration = 0; true; ++iteration) {
```
Comment (dgozman) [inline packages/playwright-core/src/server/page.ts:471]: I guess `iteration` was used at some point, but not anymore?

File: packages/playwright-test/src/matchers/toMatchSnapshot.ts (hunk 4 above)
Comment (aslushnikov) [reply packages/playwright-test/src/matchers/toMatchSnapshot.ts:299]: check for page or locator explicitly

File: packages/playwright-test/src/expect.ts (hunk 2 above)
Comment (aslushnikov) [reply packages/playwright-test/src/expect.ts]: i'll extract separately.

**Review by dgozman — APPROVED** [review]
